doc/api/esm.md

```diff
@@ -1139,11 +1139,18 @@ CommonJS modules loaded.
 
 ### Hooks
 
-#### <code>resolve</code> hook
+#### `resolve(specifier, context, defaultResolve)`
 
 > Note: The loaders API is being redesigned. This hook may disappear or its
 > signature may change. Do not rely on the API described below.
 
+* `specifier` {string}
+* `context` {Object}
+  * `conditions` {string[]}
+  * `parentURL` {string}
+* `defaultResolve` {Function}
+* Returns: {Promise}
```

Comment:
For async functions, do the Node.js API ref. docs types usually express what the promise is wrapping? Compare to the JSDoc below.

Reply:
Update: AFAIK, the Node.js API ref. docs type system lacks the ability to express what we want here. So, since this API is not strictly Promise-based (it works whether this is an async function or not), IMO this should be expressed as an Object the same way the others are.

```suggestion
* Returns: {Object}
  * `url` {string}
```

Reply:
Yet to be documented:

> Whether or not this hook should ever be written as a non-async function depends on the circumstance.


Unless others feel strongly about this or have a better suggestion, I'll take the suggestion I've proposed above. The point of this PR was that it was supposed to be as frictionless as possible, so being stuck on this seems counterproductive especially since we would want these changes made before continuing w/ https://github.com/nodejs/node/pull/33812.

Reply:
There's some additional useful information is the thread where the following was mentioned.

> if you know your value will be available synchronously don't use a promise? the entire point of promises is "value will be available at some point that I have no inference or control of and my code should still operate consistently through that"
> &mdash; https://twitter.com/devsnek/status/1284528435671465988

I'm willing to continue the discussion as it relates to this PR, so please feel free to chime if interested.